File: Taoprone/Classes/Core/TPEngine.m
//
//  TPEngine.m
//  Taoprone
//
//  Created by Rake Yang 11/12/2019.
//  Copyright (c) 2019 BinaryParadise All rights reserved.
//

#import "TPEngine.h"
#import "NSObject+TPBridge.h"

@interface TPEngine ()

@property (nonatomic, strong) JSContext *jsContext;


@end

@implementation TPEngine

- (instancetype)init {
    if (self = [super init]) {
        self.jsContext = [[JSContext alloc] initWithVirtualMachine:[JSVirtualMachine new]];
        self.jsContext.exceptionHandler = ^(JSContext *context, JSValue *exception) {
            NSLog(@"%@", exception);
        };
        //解释引擎
        NSString *enginePath = [[self.class sdkBundle] pathForResource:@"engine.js" ofType:nil];
        NSString *jsEngine = [NSString stringWithContentsOfFile:enginePath encoding:NSUTF8StringEncoding error:nil];
        [self.jsContext evaluateScript:jsEngine withSourceURL:[NSURL fileURLWithPath:enginePath]];
    }
    return self;
}

- (id)moduleWithURL:(NSString *)filePath {
    static NSString *_regexStr = @"(?<!console)\\.\\s*(\\w+)\\s*\\(";
    static NSString *_replaceStr = @".__c(\"$1\")(";
    NSRegularExpression* _regex;
    //解释引擎
    NSString *jsContent = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
    _regex = [NSRegularExpression regularExpressionWithPattern:_regexStr options:0 error:nil];
    jsContent = [_regex stringByReplacingMatchesInString:jsContent options:0 range:NSMakeRange(0, jsContent.length) withTemplate:_replaceStr];
    NSString *cachePath = [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES).firstObject stringByAppendingString:@"/Taoprone"];
    if (![[NSFileManager defaultManager] fileExistsAtPath:cachePath]) {
        [[NSFileManager defaultManager] createDirectoryAtPath:cachePath withIntermediateDirectories:YES attributes:nil error:nil];
    }

    NSString *mainJS = [cachePath stringByAppendingString:@"/main.js"];
    [jsContent writeToFile:mainJS atomically:YES encoding:NSUTF8StringEncoding error:nil];

    TPBridgeExport *bridge = [TPBridgeExport new];
    bridge.context = self.jsContext;
    self.jsContext[@"TPBridge"] = bridge;
    [self.jsContext evaluateScript:jsContent withSourceURL:[NSURL fileURLWithPath:mainJS]];
    self.jsContext[@"nslog"] = ^(NSString *msg) {
        NSLog(@"%@", msg);
    };
    return [self.jsContext[@"initModule"] callWithArguments:nil].toObject;
}

+ (NSBundle *)sdkBundle {    
    NSString *path = [[NSBundle bundleForClass:self.class] pathForResource:@"Taoprone.bundle" ofType:nil];
    return [NSBundle bundleWithPath:path];
}

- (void)dealloc {
    self.jsContext = nil;
    NSLog(@"%s+%d", __FUNCTION__, __LINE__);
}

@end

File: Taoprone/Classes/Core/TPEngine.swift
//
//  TPEngine.swift
//  Taoprone
//
//  Created by Rake Yang 11/12/2019.
//  Copyright (c) 2019 BinaryParadise All rights reserved.
//

import JavaScriptCore

public class TPEngine: NSObject {
    private var jsContext = JSContext(virtualMachine: JSVirtualMachine())
    
    public override init() {
        super.init()
        let signature = si_methodSignature(for: NSSelectorFromString("init"))
        var invocation = SIInvocation(methodSignature: signature)
        invocation.target = TPEngine.self
        invocation.invoke()
        var returnValue: Unmanaged<AnyObject>?
        invocation.getReturnValue(&returnValue)
        print("\(returnValue)")
    }
    
    public func moduleWithURL(filePath: String) throws -> UIViewController? {
        guard let jsContext = jsContext else { return nil }
        let _regexStr = "(?<!\\\\)\\.\\s*(\\w+)\\s*\\("
        let _replaceStr = ".__native(\"$1\")("
        let _regex = try NSRegularExpression(pattern: _regexStr, options: .caseInsensitive)        
        //解释引擎
        var jsEngine = try String(contentsOfFile: TPEngine.sdkBundle().path(forResource: "engine.js", ofType: nil)!)
        let jsContent = try String(contentsOfFile: filePath)
        let newContent = _regex.stringByReplacingMatches(in: jsContent, options: .reportProgress, range: NSMakeRange(0, jsContent.count), withTemplate: _replaceStr)
        let cachePath = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first!.appending("/Taoprone")
        
        if !FileManager.default.fileExists(atPath: cachePath) {
            try FileManager.default.createDirectory(atPath: cachePath, withIntermediateDirectories: true, attributes: nil);
        }

        //拼接逻辑脚本
        jsEngine = jsEngine.appending("\n//--------------------业务代码----------------------\n")
        let merged = jsEngine.appending(newContent)
        let mainJS = cachePath.appending("/main.js")
        try merged.write(toFile: mainJS, atomically: true, encoding: .utf8)

        jsContext.exceptionHandler = { context, exception in
            print("exceptionHandler: \(String(describing: exception))")
        }
        jsContext.setObject(TPBridge(), forKeyedSubscript: NSString("TPBridge"))
        
        //block封装
        let willCallClosure: @convention(block) () -> String = {
            print("Hello, world!")
            return "0.1.0 (swift 2021-06-28)"
        }
        jsContext.setObject(willCallClosure, forKeyedSubscript: NSString("sayHelloToSwift"))
        jsContext.evaluateScript(merged, withSourceURL: URL(fileURLWithPath: mainJS))
        let result = jsContext.objectForKeyedSubscript("initModule").call(withArguments: nil)
        return result?.toObject() as? UIViewController
    }

    class func sdkBundle() -> Bundle {
        let path = Bundle(for: Self.self).path(forResource: "Taoprone.bundle", ofType: nil)
        return Bundle(path: path!)!
    }
    
    deinit {
        jsContext = nil
        print("\(self) \(#function)")
    }
}
